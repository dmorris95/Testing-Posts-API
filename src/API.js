const jsonApiUrl = 'https://jsonplaceholder.typicode.com';

export const fetchPosts = async () => {
    const response = await fetch(`${jsonApiUrl}/posts`);
    if (!response.ok) {
        throw new Error('Network response was not good');
    }
    return response.json();
};

export const createPost = async (post) => {
    const response = await fetch(`${jsonApiUrl}/posts`, {
        method: 'POST',
        body: JSON.stringify(post),
        headers: {
            'Content-type': 'application/json; charset=UTF-8',
        },
    });
    if (!response.ok) {
        throw new Error('Failed to add post');
    }
    return response.json();
};

export const updatePost = async (post) => {
    const response = await fetch(`${jsonApiUrl}/posts/${post.id}`, {
        method: 'PUT',
        body: JSON.stringify(post),
        headers: {
            'Content-type': 'application/json; charset=UTF-8',
        },
    });
    if (!response.ok) {
        throw new Error('Failed to update post');
    }
    return response.json();
};

export const deletePost = async (postId) => {
    const response = await fetch(`${jsonApiUrl}/posts/${postId}`, {
        method: 'DELETE',
    });
    if (!response.ok) {
        throw new Error('Failed to delete post');
    }
    return response.json();
};