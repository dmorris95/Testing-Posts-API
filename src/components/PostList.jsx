import { useQueryClient, useMutation, useQuery } from "@tanstack/react-query";
import { useCallback, useMemo, useState } from "react";
import { Card, Button, Spinner, Alert, Form } from "react-bootstrap";
import { fetchPosts, deletePost } from "../API";
import React from "react";

const PostList = ({ onPostClick }) => {
    const queryClient = useQueryClient();
    const [filterId, setFilterId] = useState('');

    const { data, isLoading, isSuccess, error } = useQuery({
        queryKey: ['posts'],
        queryFn: fetchPosts,
        refetchOnReconnect: true,
        refetchOnWindowFocus: true,
        retry: 3,
        retryDelay: attemptIndex => Math.min(1000 * 2 ** attemptIndex, 30000),
        staleTime: 5 * 60 * 1000, 
        cacheTime: 15 * 60 * 1000
    });

    const delPost = useMutation({
        mutationFn: deletePost,
        onSuccess: (data, postId) => {
            console.log('Post deleted successfully')
            queryClient.setQueryData(['posts'], (existingPosts) => {
                return existingPosts.filter(post => post.id !== postId);
            });
        },
    });

    //Post Memoization
    const filterPosts = useMemo(() => {
        //Data check in case an undefined value is returned from the API call
        return data ? data.filter(post => post.userId.toString().includes(filterId)) : [];
    }, [data, filterId]);

    //Memoize the postClick (The UserSelect in this program)
    const handlePostClick = useCallback((post) => {
        onPostClick(post);
    }, [onPostClick]);


    if (isLoading) return <Spinner animation="border" role="status"><span className="visually-hidden">Loading...</span></Spinner>;
    if (error) return <Alert variant="danger">Error fetching data</Alert>;

    return (
        <div>
            <input className="mb-3 bg-light border-primary rounded-2" name="filterId" type="number" value={filterId} placeholder="Filter by User ID" onChange={(e) => setFilterId(e.target.value)}/>
            {isSuccess && filterPosts.map(post => (
                <Card key={post.id} className="mb-3">
                    <Card.Body>
                        <Card.Title>{post.title}</Card.Title>
                        <Card.Text>{post.body}</Card.Text>
                        <Button variant="primary" onClick={() => handlePostClick(post)}>Edit</Button>
                        <Button 
                            variant="danger" 
                            data-testid={`del-button-${post.id}`} 
                            onClick={() => delPost.mutate(post.id)}
                        >
                        Delete
                        </Button>
                    </Card.Body>
                </Card>
            ))}
        </div>
    );
}

export default PostList