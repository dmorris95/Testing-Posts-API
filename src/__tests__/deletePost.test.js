import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import '@testing-library/jest-dom';
import PostList from '../components/PostList';
import { deletePost, fetchPosts } from "../API";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

//Mock fetchPosts from the API.js
jest.mock('../API', () => ({
    deletePost: jest.fn(),
    fetchPosts: jest.fn()
}));
const queryClient = new QueryClient();
const renderUseQueryClient = (ui) => {
    return render(
        <QueryClientProvider client={queryClient}>
            {ui}
        </QueryClientProvider>
    );
};

describe('deletePost Function', () => {
    test('deleting a post via the delete button', async () => {
        const mockPosts = [
            { id: 1, title: 'Test Title 1', body: 'Post body 1', userId: 1 },
            { id: 2, title: 'Test Title 2', body: 'Post body 2', userId: 2 },
        ];

        //Simulates initial call
        fetchPosts.mockResolvedValue(mockPosts);

        //Simulate the delete call
        deletePost.mockResolvedValue({});

        const { asFragment } = renderUseQueryClient(
            <>
                <PostList onPostClick={jest.fn()} />
            </>
        );
        //First waitFor ensures posts are initialized
        await waitFor(() => {
            expect(screen.getByText('Test Title 1')).toBeInTheDocument();
            expect(screen.getByText('Test Title 2')).toBeInTheDocument();
        });

        fireEvent.click(screen.getByTestId('del-button-1'));

        //Second waitFor ensures the post was removed from the UI correctly
        await waitFor(() => {
            //QueryBy to search the document
            expect(screen.queryByText('Test Title 1')).not.toBeInTheDocument();
            expect(screen.getByText('Test Title 2')).toBeInTheDocument();
        });

        expect(deletePost).toHaveBeenCalledWith(1);
        expect(asFragment()).toMatchSnapshot();
    });
})