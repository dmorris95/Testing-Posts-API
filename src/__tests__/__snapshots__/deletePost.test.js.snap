// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deletePost Function deleting a post via the delete button 1`] = `
<DocumentFragment>
  <div>
    <input
      class="mb-3 bg-light border-primary rounded-2"
      name="filterId"
      placeholder="Filter by User ID"
      type="number"
      value=""
    />
    <div
      class="mb-3 card"
    >
      <div
        class="card-body"
      >
        <div
          class="card-title h5"
        >
          Test Title 2
        </div>
        <p
          class="card-text"
        >
          Post body 2
        </p>
        <button
          class="btn btn-primary"
          type="button"
        >
          Edit
        </button>
        <button
          class="btn btn-danger"
          data-testid="del-button-2"
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
